lib/collection/src/collection.rs

```diff
@@ -258,17 +258,79 @@ impl Collection {
         todo!("Handle for sender and receiver")
     }
 
+    pub async fn initiate_temporary_shard(&self, shard_id: ShardId) -> CollectionResult<()> {
+        let shard_holder_read = self.shards_holder.read().await;
+        // validate that the shard is known as a remote shard
+        if let Some(Shard::Remote(_)) = shard_holder_read.get_shard(&shard_id) {
+            // do not lock shards while creating temporary shard
+            drop(shard_holder_read);
+            let shard_path = Self::create_temporary_shard_dir(&self.path, shard_id).await?;
+            let temporary_shard =
+                LocalShard::build(shard_id, self.id.clone(), &shard_path, self.config.clone())
+                    .await?;
+            // short write lock to insert the temporary shard
+            let mut shards_holder_write = self.shards_holder.write().await;
+            shards_holder_write.add_temporary_shard(shard_id, temporary_shard);
+            Ok(())
+        } else {
+            Err(CollectionError::BadRequest {
+                description: "A shard transfer must target an existing remote shard".to_string(),
+            })
+        }
+    }
+
+    /// Create directory to hold the temporary data for shard with `shard_id`
+    pub async fn create_temporary_shard_dir(
+        collection_path: &Path,
+        shard_id: ShardId,
+    ) -> CollectionResult<PathBuf> {
+        let shard_path = collection_path.join(format!("{shard_id}-temp"));
+        tokio::fs::create_dir_all(&shard_path)
+            .await
+            .map_err(|err| CollectionError::ServiceError {
+                error: format!(
+                    "Can't create shard {shard_id} temporary directory. Error: {}",
+                    err
+                ),
+            })?;
+        Ok(shard_path)
+    }
+
+    /// Handle collection updates from peers.
+    ///
+    /// Shard transfer aware.
     pub async fn update_from_peer(
         &self,
         operation: CollectionUpdateOperations,
         shard_selection: ShardId,
         wait: bool,
     ) -> CollectionResult<UpdateResult> {
-        let local_shard = self
-            .shards_holder
-            .local_shard_by_id(shard_selection)
-            .await?;
-        local_shard.get().update(operation.clone(), wait).await
+        let shard_holder_guard = self.shards_holder.read().await;
+        let shard_opt = shard_holder_guard.get_shard(&shard_selection);
+
+        let target_shard = match shard_opt {
+            None => {
+                return Err(CollectionError::bad_shard_selection(format!(
```

Comment:
Theoretically, it might be possible that we create a tmp shard for which there are no Remote shard exists just yet. e.g. in case of re-sharding. Let's handle this case right away.

Reply:
https://github.com/qdrant/qdrant/pull/829/commits/f75252370c1dd2651a0daac66e0487fe6bcba39e